### PR TITLE
[Promotion] Apply promotion every Nth Product

### PR DESCRIPTION
**Pull request by brunoroux**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

This promotion should apply on product quantity (not variation).

Meaning : 

Given I have 2 products : "PHP T-shirt" and "Sylius T-shirt"
"PHP T-shirt" has 2 variations "Size L" and "Size XL", price is : 20€
"Sylius T-shirt" doesn't have variations, price is : 20€

My Promotion has the rule "Nth product : 2"
My Promotion has the action "Nth product fixed discount amount : 5€ nth : 2"

If I add 2 "PHP T-shirt Size L", my promotion should be 5 €, my cart total should be 35 €
If I add 1 "PHP T-shirt Size L" and 1 "PHP T-shirt Size XL", my promotion should be 5€, my cart total should be 35 €
If I add 1 "PHP T-shirt" and 1 "Sylius T-shirt", I shouldn't have any promotion
If I add 3 "PHP T-shirt Size L" and 1 "PHP T-shirt XL" my promotion should be 10€, my cart total should be 70 €

Note : This could probably be improved once https://github.com/Sylius/Sylius/issues/1810 is done

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Form/Type/Rule/NthProductConfigurationType.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Sylius\Bundle\CoreBundle\Form\Type\Rule;
+
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Validator\Constraints\NotBlank;
+use Symfony\Component\Validator\Constraints\Type;
+
+
```
Comment (aramalipoor) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Rule/NthProductConfigurationType.php]: Missing you as author and an extra line.

File: src/Sylius/Bundle/CoreBundle/Form/Type/Rule/NthProductConfigurationType.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
+<?php
+
```
Comment (aramalipoor) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Rule/NthProductConfigurationType.php]: Missing copyright.

File: src/Sylius/Component/Core/Promotion/Checker/NthProductRuleChecker.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ */
+class NthProductRuleChecker implements RuleCheckerInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function isEligible(PromotionSubjectInterface $subject, array $configuration)
+    {
+        if (!$subject instanceof OrderInterface) {
+            throw new UnsupportedTypeException($subject, 'Sylius\Component\Core\Model\OrderItemInterface');
```
Comment (aramalipoor) [inline src/Sylius/Component/Core/Promotion/Checker/NthProductRuleChecker.php]: `OrderItemInterface` or `OrderInterface`?

File: src/Sylius/Component/Core/Promotion/Checker/NthProductRuleChecker.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        }
+
+        foreach ($subject->getItems() as $item) {
+            if ($item->getQuantity() >= $configuration['nth']) {
+                return true;
+            }
```
Comment (aramalipoor) [inline src/Sylius/Component/Core/Promotion/Checker/NthProductRuleChecker.php]: I didn't quite get the logic here. If any of the items has a quantity more than what's configured the rule applies?

So if I order 10 different products with 1 quantity I'm not going to get a promotion of `nth promotion with 2 as config`?

File: src/Sylius/Bundle/CoreBundle/Form/Type/Rule/NthProductConfigurationType.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                'constraints' => array(
+                    new NotBlank(),
+                    new Type(array('type' => 'numeric')),
+                )
+            ))
+        ;
+    }
+
+    /**
+     * Configures the options for this type.
```
Comment (aramalipoor) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Rule/NthProductConfigurationType.php]: Redundant comment.

File: src/Sylius/Component/Core/Promotion/Checker/NthProductRuleChecker.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
+<?php
+
```
Comment (aramalipoor) [inline src/Sylius/Component/Core/Promotion/Checker/NthProductRuleChecker.php]: Missing copyright

**Comment by aramalipoor** [comment]
Depending the logic I see, maybe `OrderItemQuantityRule` is a better name IMO.

File: src/Sylius/Component/Core/Promotion/Checker/NthProductRuleChecker.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+namespace Sylius\Component\Core\Promotion\Checker;
+
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Promotion\Checker\RuleCheckerInterface;
+use Sylius\Component\Promotion\Exception\UnsupportedTypeException;
+use Sylius\Component\Promotion\Model\PromotionSubjectInterface;
+use Symfony\Component\VarDumper\VarDumper;
+
+/**
+ * checks if order contains nth times the same product
```
Comment (aramalipoor) [inline src/Sylius/Component/Core/Promotion/Checker/NthProductRuleChecker.php]: Add yourself as author ;)

File: src/Sylius/Bundle/CoreBundle/Form/Type/Rule/NthProductConfigurationType.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Class NthProductConfigurationType
+ *
+ * @author Bruno Roux <bruno@yproximite.com>
+ */
+class NthProductConfigurationType extends AbstractType
+{
+    protected $validationGroups;
+
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Rule/NthProductConfigurationType.php]: Missing docblock.

File: features/backend/promotions.feature [hunk 9]
```diff
@@ -87,6 +87,18 @@ Feature: Promotions
           And I should see "Promotion has been successfully created."
 
     @javascript
+    Scenario: Creating new promotion with nth item rule
+        Given I am on the promotion creation page
+         When I fill in "Name" with "Behat Training"
+          And I fill in "Description" with "Behat Training discount on 2nd same item"
+          And I click "Add rule"
+          And I select "Nth item" from "Type"
```
Comment (pamil) [inline features/backend/promotions.feature]: Should be "Nth product"

**Pull request closed without merging**